cmake/SetupCxxFlags.cmake

```diff
@@ -69,6 +69,18 @@ set_property(TARGET SpectreFlags
   $<$<COMPILE_LANGUAGE:CXX>:-fno-math-errno>
   $<$<COMPILE_LANGUAGE:Fortran>:-fno-math-errno>)
 
+# Allow compiler to combine (x * y + z) into an FMA.
+#
+# Note: this can be problematic in some edge cases where a norm can become
+# negative due to the different rounding properties of FMA so if we run into
+# problems we'll need to reassess.
+set_property(TARGET SpectreFlags
+  APPEND PROPERTY
+  INTERFACE_COMPILE_OPTIONS
+  $<$<COMPILE_LANGUAGE:C>:-ffp-contract=on>
+  $<$<COMPILE_LANGUAGE:CXX>:-ffp-contract=on>
+  $<$<COMPILE_LANGUAGE:Fortran>:-ffp-contract=on>)
```

Comment:
I looked into this one in a bit more detail, and it looks like this option means different things in clang and gcc.  For clang "on" is the default, and for GCC it looks like "on" actually turns off converting things to FMA instructions.  So we should  probably leave this one out.

GCC 11.3.1:
>        -ffp-contract=style
>            -ffp-contract=off disables floating-point expression contraction.
>            -ffp-contract=fast enables floating-point expression contraction
>            such as forming of fused multiply-add operations if the target has
>            native support for them.  -ffp-contract=on enables floating-point
>            expression contraction if allowed by the language standard.  This
>            is currently not implemented and treated equal to
>            -ffp-contract=off.
> 
>            The default is -ffp-contract=fast.

Clang (whatever version is documented on their website):
> -ffp-contract=\<arg>
>
> Form fused FP ops (e.g. FMAs): fast (fuses across statements disregarding pragmas) | on (only fuses in the same statement unless dictated by pragmas) | off (never fuses) | fast-honor-pragmas (fuses across statements unless dictated by pragmas). Default is ‘fast’ for CUDA, ‘fast-honor-pragmas’ for HIP, and ‘on’ otherwise. <arg> must be ‘fast’, ‘on’, ‘off’ or ‘fast-honor-pragmas’.

Reply:
Sigh... If only things were more consistent. I agree, let's omit. If we need to change to `fast-honor-pragmas` for clang later we can!